### PR TITLE
Support unmarshaling of SX1301 Conf Messages

### LINKED ISSUE #808: Support Unmarshalling of SX1301 Conf
#### Summary

Support Unmarshalling of SX1301 Conf

#### Why do we need this?

To un-marshal SX1301 Configuration when used stand-alone or as part of packet-forwarder/router config.

#### What is already there? What do you see now?

Not Implemented `errUnmarshalNotImplemented `

#### What is missing? What do you want to see?

Implement it.

#### Environment

`ttn-lw-stack v3.03`

#### How do you propose to implement this?

- Multi-stage un-marshaling due to the sub-messages.

#### Can you do this yourself and submit a Pull Request?

Yup

### PR DESCRIPTION
**Pull request by KrishnaIyer**
#### Summary

Closes #808 

#### Changes

- Implement unmarshaling.
- Add corresponding test to Basic Station.

#### Notes for Reviewers

- As discussed in the issue, this is only to support tests.

#### Release Notes

- Added unmarshaling support to SX1301Conf.

### CODE REVIEW

File: pkg/pfconfig/shared/shared.go [hunk 1]
```diff
@@ -108,10 +107,90 @@ func (c SX1301Config) MarshalJSON() ([]byte, error) {
 	return json.Marshal(m)
 }
 
-// UnmarshalJSON implements json.Unmarshaler. It currently just errors because
-// unmarshaling is not supported.
-func (c *SX1301Config) UnmarshalJSON(_ []byte) error {
-	return errUnmarshalNotImplemented
+var (
+	radioKey = regexp.MustCompile("^radio_([0|1])$")
```
Comment (htdvisser) [inline pkg/pfconfig/shared/shared.go]: I don't think you really meant `([0|1])` here. Likely `([01])` or `(0|1)`.

File: pkg/pfconfig/shared/shared.go [hunk 2]
```diff
@@ -108,10 +107,90 @@ func (c SX1301Config) MarshalJSON() ([]byte, error) {
...
+	chanKey  = regexp.MustCompile("^chan_multiSF_([0-9]{1,2})$")
+)
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (c *SX1301Config) UnmarshalJSON(msg []byte) error {
+	var root map[string]json.RawMessage
+	if err := json.Unmarshal(msg, &root); err != nil {
+		return err
+	}
+	radioMap, txLutMap, chanMap := make(map[int]RFConfig), make(map[int]TxLUTConfig), make(map[int]IFConfig)
```
Comment (htdvisser) [inline pkg/pfconfig/shared/shared.go]: Nit: don't really like this style

File: pkg/pfconfig/shared/shared.go [hunk 3]
```diff
@@ -108,10 +107,90 @@ func (c SX1301Config) MarshalJSON() ([]byte, error) {
...
+	for key, value := range root {
+		if key == "lorawan_public" {
```
Comment (htdvisser) [inline pkg/pfconfig/shared/shared.go]: Better to use a `switch` statement here, no need to do all comparisons for every key.

```go
switch {
	case key == "lorawan_public":
	// ...
}
```

File: pkg/pfconfig/shared/shared.go [hunk 4]
```diff
@@ -108,10 +107,90 @@ func (c SX1301Config) MarshalJSON() ([]byte, error) {
...
+			if err := json.Unmarshal(value, &c.LoRaStandardChannel); err != nil {
+				return err
+			}
+		}
+		if key == "chan_FSK" {
+			if err := json.Unmarshal(value, &c.FSKChannel); err != nil {
+				return err
+			}
+		}
+		if match := radioKey.FindStringSubmatch(key); len(match) == 2 {
```
Comment (htdvisser) [inline pkg/pfconfig/shared/shared.go]: Nit: Using a `strings.HasPrefix` and `fmt.Scanf` is probably more efficient than matching 3 regexes in order to get to the most common one (the IF config). Could also choose to re-order.

File: pkg/pfconfig/shared/shared.go [hunk 5]
```diff
@@ -108,10 +107,90 @@ func (c SX1301Config) MarshalJSON() ([]byte, error) {
...
+			if err := json.Unmarshal(value, &channel); err != nil {
+				return err
+			}
+			if index, err := strconv.Atoi(match[1]); err == nil {
+				chanMap[index] = channel
+			}
+		}
+	}
+
+	c.Radios, c.TxLUTConfigs, c.Channels = make([]RFConfig, len(radioMap)), make([]TxLUTConfig, len(txLutMap)), make([]IFConfig, len(chanMap))
```
Comment (htdvisser) [inline pkg/pfconfig/shared/shared.go]: Nit: don't really like this style

File: pkg/pfconfig/shared/shared.go (hunk 4 above)
Comment (KrishnaIyer) [reply pkg/pfconfig/shared/shared.go]: Damn this is super cool. I don't need those regexes. Also the only reason for using the `if` was the regex match.
Thanks!

File: pkg/pfconfig/shared/shared.go [hunk 6]
```diff
@@ -108,10 +105,80 @@ func (c SX1301Config) MarshalJSON() ([]byte, error) {
...
+			if err := json.Unmarshal(value, &c.FSKChannel); err != nil {
+				return err
+			}
+		case strings.HasPrefix(key, "chan_multiSF_"):
+			var channel IFConfig
+			if err := json.Unmarshal(value, &channel); err != nil {
+				return err
+			}
+			var index int
+			if _, err := fmt.Sscanf(key, "chan_multiSF_%d", &index); err == nil {
```
Comment (htdvisser) [inline pkg/pfconfig/shared/shared.go]: We could be a bit more strict and return this `err` if it's non-nil

Comment (KrishnaIyer) [reply pkg/pfconfig/shared/shared.go]: ok done. Re-requesting review since it won't let me merge. 🙏

**Comment by KrishnaIyer** [comment]
@johanstokking: Seems like you're the code owner for this and requires your approval.

**Pull request merged**